ckanext/unhcr/blueprints/kobo.py

```
@@ -0,0 +1,123 @@
+        return toolkit.redirect_to('unhcr_kobo.index')
+    plugin_extras = {} if user_obj.plugin_extras is None else user_obj.plugin_extras
+    old_token = plugin_extras.get('unhcr', {}).get('kobo_token')
+
+    if token != old_token:
+        try:
+            toolkit.get_action('user_update')(
+                {'user': user_obj.name, 'validate_token': not token.startswith('test_')},
+                {'id': user_obj.id, 'plugin_extras': {'unhcr':  {'kobo_token': token}}}
+            )
+        except toolkit.ValidationError as e:
+            if e.error_dict and 'kobo_token' in e.error_dict:
+                message = 'Invalid KoBo token. Fix it and try again.'
+            else:
+                message = 'Unknown KoBo token error: {}.'.format(e)
+            toolkit.h.flash_error(message)
+            return toolkit.redirect_to('unhcr_kobo.index')
+
+    return toolkit.redirect_to(redirect_on_success)
+
+
+def remove_token(redirect_on_success='unhcr_kobo.index'):
+    """ POST a new token """
+    user_obj = toolkit.c.userobj
+
+    try:
+        toolkit.get_action('user_update')(
+            {'user': user_obj.name},
+            {'id': user_obj.id, 'plugin_extras': {'unhcr':  {'kobo_token': 'REMOVE'}}}
+        )
+    except toolkit.ValidationError as e:
+        message = 'Error removing token: {}.'.format(e)
+        toolkit.h.flash_error(message)
+        return toolkit.redirect_to('unhcr_kobo.index')
+
+    return toolkit.redirect_to(redirect_on_success)
+
+
+@require_editor_user
+def kobo_surveys():
+    """ KoBo surveys page """
+    user_obj = toolkit.c.userobj
+
+    plugin_extras = {} if user_obj.plugin_extras is None else user_obj.plugin_extras
+    token = plugin_extras.get('unhcr', {}).get('kobo_token')
+
+    # check if the token is valid
+    kobo_url = toolkit.config.get('ckanext.unhcr.kobo_url', 'https://kobo.unhcr.org')
+    kobo = KoBoAPI(token, kobo_url)
+
+    try:
+        surveys = kobo.get_surveys()
+    except Exception as e:
```

We really need to avoid catch-all Exceptions like this one. I suggest creating a `KoboException` class in the `kobo/api.py` module that we raise explicitly when needed (ie just [here](https://github.com/okfn/ckanext-unhcr/pull/635/files#diff-7177da887f4e624e91e757a37fef4883258d73484fe9a58a4fb0639d19debe5eR25) right now) and catching it here instead of `Exception`. Otherwise we can mask other unrelated failures.

---

Added Kobo error [here](https://github.com/okfn/ckanext-unhcr/pull/635/commits/002a3b5bfedcb3af4fbdbd0ce0498eed2e2f29b9).